compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp

```diff
@@ -0,0 +1,537 @@
+         "expected as many tile sizes as loop ranges");
+  assert(loopRanges.size() == distributionMethod.size() &&
+         "expected as many entries in distribution method list as number of "
+         "loop ranges");
+  OpBuilder::InsertionGuard guard(builder);
+  SmallVector<scf::ForOp> loops;
+  offsets.resize(loopRanges.size());
+  sizes.resize(loopRanges.size());
+
+  // The tile size to use (to avoid out of bounds access) is  minimum of
+  // `tileSize` and `ub - iv`, where `iv` is the induction variable
+  // of the tiled loop.
+  AffineExpr s0, s1, d0;
+  bindDims(builder.getContext(), d0);
+  bindSymbols(builder.getContext(), s0, s1);
+  AffineMap minMap = AffineMap::get(1, 2, {s0, s1 - d0}, builder.getContext());
+  auto createBoundedTileSize = [&](Value iv, Value tileSize,
+                                   Value size) -> OpFoldResult {
+    if (isConstantIntValue(getAsOpFoldResult(tileSize), 1)) {
+      return builder.getIndexAttr(1);
+    }
+    return makeComposedFoldedAffineMin(
+        builder, loc, minMap, ArrayRef<OpFoldResult>{iv, tileSize, size});
+  };
+
+  unsigned procDim = 0;
+  for (auto loopRange : llvm::enumerate(loopRanges)) {
+    Value lb = loopRange.value().offset;
+    Value ub = loopRange.value().size;
+    Value step = tileSizeVals[loopRange.index()];
+
+    // No loops if tile size is zero. Set offset and size to the loop
+    // offset and size.
+    if (matchPattern(tileSizeVals[loopRange.index()], m_Zero())) {
+      offsets[loopRange.index()] = lb;
+      sizes[loopRange.index()] = ub;
+      continue;
+    }
+
+    auto method = distributionMethod[loopRange.index()];
+    if (method != linalg::DistributionMethod::None) {
+      std::tie(lb, step) = getDistributeLBAndStep(builder, loc, lb, step,
+                                                  procInfo[procDim].procId,
+                                                  procInfo[procDim].nprocs);
+      procDim++;
+    }
+
+    if (method == linalg::DistributionMethod::CyclicNumProcsEqNumIters) {
+      offsets[loopRange.index()] = getAsOpFoldResult(lb);
+      OpFoldResult boundedTileSize =
+          createBoundedTileSize(lb, tileSizeVals[loopRange.index()], ub);
+      sizes[loopRange.index()] = boundedTileSize;
+      continue;
+    }
+
+    auto loop = builder.create<scf::ForOp>(
+        loc, lb, ub, step, ValueRange{},
+        [&](OpBuilder &bodyBuilder, Location bodyLoc, Value iv,
+            ValueRange /*iterArgs*/) {
+          OpFoldResult boundedTileSize =
+              createBoundedTileSize(iv, tileSizeVals[loopRange.index()], ub);
+          sizes[loopRange.index()] = boundedTileSize;
+          builder.create<scf::YieldOp>(loc);
+        });
+    offsets[loopRange.index()] = loop.getInductionVar();
+    loops.push_back(loop);
+    builder.setInsertionPoint(loop.getBody()->getTerminator());
+  }
+  return loops;
+}
+
+namespace {
+
+/// Result of the tiled operation.
+struct TilingResult {
```

Comment:
This can be replaced by upstream `SCFTilingResult`?

Reply:
Yeah this entire code is essentially adapted from what is upstream. I'd like to keep a clean separation to not have some wired coupling. So its just an extra struct, that I just decided to duplicate.